src/chains/ethereum/ethereum/src/block-time.ts

```diff
@@ -0,0 +1,97 @@
+export interface BlockTime {
+  /**
+   * Set the current time of the BlockTime instance. Mining immediately afterwards will result in a block timestamp of ~timestamp (some milliseconds could pass before the block is mined).
+   * @param  {number | Date} timestamp - The time that should be considered "current time".
+   * @returns number - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setTime(timestamp: number | Date): number;
+  
+  /**
+   * Get the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @returns number - The "offset" for the BlockTime instance in milliseconds.
+   */
+  getOffset(): number;
+  
+  /**
+   * Set the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @param {number} offset - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setOffset(offset: number): void;
+
+  /**
+   * Create a timestamp for a new block in seconds. Can specify the timestamp for the block, in milliseconds, which will be reflected in future calls to this function.
+   * @param  {number} timestamp? - Optional - The timestamp for the block that will be mined, in milliseconds.
+   * @returns number - the timestamp in seconds
+   */
+  createBlockTimestampInSeconds(timestamp?: number | Date): number;
+}
+
+export class ClockBasedBlockTime implements BlockTime {
+  #getReferenceClockTime: () => number | Date;
+  #timeOffset: number;
+
+  private static validateTimestamp(timestamp: number | Date) {
+    if (timestamp < 0) {
+      throw new Error(`Invalid timestamp: ${timestamp}. Value must be positive.`);
```

Comment:
As discussed separately, although a negative timestamp is valid, the block `timestamp` property must be a positive integer value (it is of type `Quantity`). Presently Ganache will accept a negative value (programatically, but not via cli), but will result in undesired block time. 

I've added a comment explaining why we are validating. I don't think this should be considered a breaking change, as it's aligning Ganache with the expected behaviour, _but_ it could cause problems for our users, so I'm happy to push that change out to milestone 8 if desired.